### PR TITLE
feat(hovercard): add hovercard implementation

### PR DESCRIPTION
**Pull request by ninamarina**
# feat(hovercard): add hovercard implementation

## :hammer_and_wrench: Type Of Change

These types will increment the version number on release:

- [ ] Fix
- [x] Feature
- [ ] Performance Improvement
- [ ] Refactor

These types will not increment the version number, but will still deploy to documentation site on release:

- [ ] Documentation
- [ ] Build system
- [ ] CI
- [ ] Style (code style changes, not css changes)
- [ ] Test
- [ ] Other (chore)

## :book: Jira Ticket
https://dialpad.atlassian.net/browse/DLT-616

## :book: Description
Adds story following figma example.
Implements delay for when multiple hovercards are present. To do this, a singleton is used to create a global timer that keeps track of the id of the current hovercard being displayed, and changes it according to the logic that was defined in https://dialpad.atlassian.net/browse/DLT-1144
Adds tests.

## :pencil: Checklist

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [x] I have reviewed my changes.
- [x] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

For all Vue changes:

- [x] I have added / updated unit tests.
- [x] I have made my changes in Vue 2 and Vue 3. Note: you may sync your changes from Vue 2 to Vue 3 (or vice versa) using the `./scripts/dialtone-vue-sync.sh` script.
- [ ] I have validated components with a screen reader.
- [x] I have validated components keyboard navigation.

For all CSS changes:

- [ ] I have used design tokens whenever possible.
- [ ] I have considered how this change will behave on different screen sizes.
- [ ] I have visually validated my change in light and dark mode.
- [ ] I have used gap or flexbox properties for layout instead of margin whenever possible.

If new component:

- I am exporting any new components or constants:
  - [x] from the index.js in the component directory.
  - [x] from the index.js in the root (either `packages/dialtone-vue2` or `packages/dialtone-vue3`).
- [ ] I have added the styles for the new component to the `packages/dialtone-css` package.
- [ ] I have created a page for the new component on the documentation site in `apps/dialtone-documentation`.
- [x] I have added the new component to `common/components_list.cjs` in `packages/dialtone-vue2` and `packages/dialtone-vue3`
- [x] I have created a component story in storybook
- [x] I have created story / stories for any relevant component variants in storybook
- [x] I have created a docs page for the component in storybook.
- [x] I have checked that changing all props/slots via the UI in storybook works as expected.

## :camera: Screenshots / GIFs

hovercards

### CODE REVIEW

File: packages/dialtone-vue2/components/hovercard/hovercard.stories.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  Elementum fames nullam elementum velit proin vitae aliquet.
+  Platea nulla consectetur consequat sagittis nullam et ultricies nisl rhoncus
+  aliquet elementum venenatis quisque.</p>`,
+  },
+  {
+    name: 'Purdie Afra', time: '5:16 PM', default: `<p class="d-fs-200 d-lh-400 d-wmx90ch">
+  Elementum fames nullam elementum velit proin vitae aliquet.
+  Platea nulla consectetur consequat sagittis nullam et ultricies nisl rhoncus
+  aliquet elementum venenatis quisque.</p>`,
+  }],
```
Comment (braddialpad) [inline packages/dialtone-vue2/components/hovercard/hovercard.stories.js]: Because this data is not part of the component interface itself we don't want it to show up in the storybook controls. We can either manually hide it from the controls via argTypesData, or rather than passing it through args put it directly in the `.story.vue`. (or in a constants file if used in multiple places).

File: packages/dialtone-vue3/components/popover/popover.test.js [hunk 2]
```diff
@@ -92,7 +92,7 @@ describe.skip('DtPopover Tests', () => {
         expect(popoverWindow.exists()).toBe(true);
       });
 
-      it('should render the main content', () => {
+      it.skip('should render the main content', () => {
```
Comment (braddialpad) [inline packages/dialtone-vue3/components/popover/popover.test.js:95]: I think perhaps the same problem you're having with your hovercard tests, is the same reason these tests are skipped.

File: packages/dialtone-vue3/components/popover/popover.vue [hunk 3]
```diff
@@ -521,6 +522,29 @@ export default {
...
+    },
+
+    /**
+     * A prefix for the data-qa tags in the elements inside the popover.
+     * If null, `dt-popover` will be used.
+     */
+    qa: {
+      type: String,
+      default: null,
+    },
```
Comment (braddialpad) [inline packages/dialtone-vue3/components/popover/popover.vue]: Can we just pass a standard `data-qa="value"` from the hovercard and then do something like 
```
:data-qa="$attrs['data-qa'] ? `${ $attrs['data-qa'] }-content` : 'dt-popover-content'"
``` 
and so on? The prop feels a little awkward to me but I'm not totally against it if you  want to do it that way.

**Review by braddialpad — COMMENTED** [review]
Couple quick comments, will look more thoroughly tomorrow.

Examples look great! Delay feels great IMO.

File: packages/dialtone-vue3/components/popover/popover.vue (hunk 3 above)
Comment (ninamarina) [reply packages/dialtone-vue3/components/popover/popover.vue]: I tried that, but the attrs where not passed to the child component. Will investigate more if that can be possible.

File: packages/dialtone-vue3/components/popover/popover.test.js (hunk 2 above)
Comment (ninamarina) [reply packages/dialtone-vue3/components/popover/popover.test.js:95]: Might be, although in vue 2 they are not skipped. Might have something to do with the fact that the hovercard adds another wrapper to all this?

File: packages/dialtone-vue3/components/popover/popover.vue (hunk 3 above)
Comment (ninamarina) [reply packages/dialtone-vue3/components/popover/popover.vue]: That was fixed in the latest changes :)

**Comment by braddialpad** [comment]
Fixes look good thanks! All that's left are these tests...

File: packages/dialtone-vue2/components/hovercard/hovercard.mdx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+<Canvas>
+  <Story of={HovercardStories.Default} />
+</Canvas>
+
+## When there are many hovercards
+
+Initially, there is a 300ms delay when numerous hovercards are present.
+However, when transitioning from one anchor to another, the delay is determined by the mouse movement time.
+This implies that the closer the anchors are, the faster the hovercard appears.
+Move the mouse over the avatars to reveal the hovercards in the example.
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard.mdx]: Should we explain here that is dependant of a global timer that need to be set it up?

File: packages/dialtone-vue2/components/hovercard/hovercard.mdx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+</Canvas>
+
+## Slots, Props & Events
+
+<Controls />
+
+## Accessibility
+
+
+## Usage
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard.mdx]: Delete empty sections or add it information

File: packages/dialtone-vue2/components/hovercard/hovercard.test.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+const baseScopedSlots = {
+  anchor: '<template #anchor="{ attrs }">' +
+          '<button data-qa="dt-button" v-bind="attrs">Hover me</button>' +
+          '</template>',
+};
+
+let mockProps = {};
+let mockAttrs = {};
+let mockSlots = {};
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard.test.js]: these are not being used, it could be removed, up to you

File: packages/dialtone-vue2/components/hovercard/hovercard.vue [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+    /**
+   * The direction the popover displays relative to the anchor.
+   * <a
+   *   class="d-link"
+   *   href="https://atomiks.github.io/tippyjs/v6/all-props/#placement"
+   *   target="_blank"
+   * >
+   *   Tippy.js docs
+   * </a>
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard.vue]: This is a styled doc for our storybook, I will suggest to move all this links to the `.mdx` file and have here the short description for the component.

File: packages/dialtone-vue2/components/hovercard/hovercard.vue [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+   * @values top, top-start, top-end,
+   * right, right-start, right-end,
+   * left, left-start, left-end,
+   * bottom, bottom-start, bottom-end,
+   * auto, auto-start, auto-end
+   */
+    placement: {
+      type: String,
+      default: 'top-start',
+    },
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard.vue]: Since we have validator on padding prop, we could have a validator here, up to you

File: packages/dialtone-vue2/components/hovercard/hovercard_constants.js
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard_constants.js:1]: If it is not necessary to have this file we could remove it.

File: packages/dialtone-vue2/components/hovercard/hovercard_default.story.vue [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+</template>
+
+<script setup>
+import DtHovercard from './hovercard.vue';
+import DtRecipeContactRow from '@/recipes/leftbar/contact_row/contact_row.vue';
+import defaultImage from '@/common/assets/avatar2.png';
+import DtStack from '../stack/stack.vue';
+import DtIcon from '../icon/icon.vue';
+import DtButton from '../button/button.vue';
+import DtAvatar from '../avatar/avatar.vue';
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard_default.story.vue:121]:
```
import { DtStack } from '@/components/stack';
import { DtIcon } from '@/components/icon';
import { DtButton } from '@/components/button';
import { DtAvatar } from '@/components/avatar';
```

File: packages/dialtone-vue2/components/hovercard/hovercard_many.story.vue [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+<script setup>
+import DtHovercard from './hovercard.vue';
+import DtRecipeFeedItemRow from '@/recipes/conversation_view/feed_item_row/feed_item_row.vue';
+import DtRecipeEmojiRow from '@/recipes/conversation_view/emoji_row/emoji_row.vue';
+import avatar1 from '@/common/assets/avatar1.png';
+import avatar2 from '@/common/assets/avatar2.png';
+import DtStack from '../stack/stack.vue';
+import DtIcon from '../icon/icon.vue';
+import DtButton from '../button/button.vue';
+import DtAvatar from '../avatar/avatar.vue';
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard_many.story.vue:145]:
```
import { DtStack } from '@/components/stack';
import { DtIcon } from '@/components/icon';
import { DtButton } from '@/components/button';
import { DtAvatar } from '@/components/avatar';
```

File: packages/dialtone-vue2/components/hovercard/hovercard.vue (hunk 7 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.vue]: Do you have an example of a component where this is done? I see the docs are filled automatically from these comments

File: packages/dialtone-vue2/components/popover/popover.vue [hunk 11]
```diff
@@ -917,6 +962,56 @@ export default {
...
+    },
+
+    setOutTimer () {
+      this.outTimer = setTimeout(() => {
+        this.isOpen = false;
+      }, TOOLTIP_DELAY_MS);
+    },
+
+    onEnterAnchor () {
+      if (!this.hovercard) return;
```
Comment (iropolo) [inline packages/dialtone-vue2/components/popover/popover.vue:984]: I would suggest to have a `isHovercard` computed property to improve readability, instead to use `this.hovercard`

File: packages/dialtone-vue2/components/hovercard/hovercard.test.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+  beforeEach(() => {
+    vi.useFakeTimers();
+    vi.runAllTimers();
+    updateWrapper();
+  });
+
+  afterEach(() => {
+    vi.useRealTimers();
+    vi.clearAllTimers();
```
Comment (iropolo) [inline packages/dialtone-vue2/components/hovercard/hovercard.test.js:60]: Not so sure about using timers in this tests....
If you remove all of them the tests still passes, It would be necessary to debug this a little bit to see if it indeed affecting something. 
If it is not, maybe it is trowing a fake check.

File: packages/dialtone-vue2/components/hovercard/hovercard.vue (hunk 7 above)
Comment (iropolo) [reply packages/dialtone-vue2/components/hovercard/hovercard.vue]: I mean, we are affecting the component prop section for storybook purpose.

Thats the point of adding stuff like this:
 ```
* <a
   *   class="d-link"
   *   href="https://popper.js.org/docs/v2/modifiers/flip/#fallbackplacements"
   *   target="_blank"
   * >
   *   Popper.js docs
   * </a>
   * */
```
Screenshot 2024-01-11 at 15 18 35

As you can see I just point it out there is a link where the consumer can check whatever they want related the prop usage. But if this addition also does not add value I will just skip it. (For example if it is does not explain something else that we are not explaining in our doc).

What I suggest is to do something like this:

image


Then we avoid to create noise in component prop description.

It is just a suggestion it is not necessary to be done on this way, but just heads up we are interfering component props doc for storybook purposes.

File: packages/dialtone-vue2/components/lazy_show/lazy_show.vue [hunk 13]
```diff
@@ -11,7 +11,7 @@
       v-on="$listeners"
     >
       <!-- @slot Slot for main content -->
-      <slot v-if="initialized" />
```
Comment (braddialpad) [inline packages/dialtone-vue2/components/lazy_show/lazy_show.vue]: Removing this makes the component not perform lazy show anymore. the slot content will be rendered immediately rather than rendering on first time it is displayed. That said there is probably a better way of doing this. In vue 3 maybe we can use suspense? https://vuejs.org/guide/built-ins/suspense.html

File: packages/dialtone-vue2/components/hovercard/hovercard.vue (hunk 7 above)
Comment (braddialpad) [reply packages/dialtone-vue2/components/hovercard/hovercard.vue]: I'm not sure we've been super strict on this historically, but it does make sense to make our in component comments pure jsdoc.. maybe something like this

```
/**
@see https://atomiks.github.io/tippyjs/v6/all-props/#placement
@values top, top-start, ...
```

and then in the stories.js override the description with the markup shown here in argTypes

File: packages/dialtone-vue2/components/hovercard/hovercard.vue (hunk 8 above)
Comment (braddialpad) [reply packages/dialtone-vue2/components/hovercard/hovercard.vue]: Tooltip has this validator, but popover doesn't for some reason... can probably just copy it from tooltip.

File: packages/dialtone-vue2/components/hovercard/hovercard.test.js (hunk 12 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.test.js:60]: This is quite weird, without the timers, the tests do pass, but the popover `isOpen` is never set. So it's giving a false positive.
With the timers, this log does show:
image

Will investigate, thanks for catching this!

File: packages/dialtone-vue3/components/hovercard/timer.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
+import { ref } from 'vue';
+import { TOOLTIP_DELAY_MS } from '@/components/tooltip/index.js';
+
+export default (function () {
+  const current = ref(null);
+  let timerLeave, timerEnter;
+  let start = null;
+  let prevCard = null;
+
+  function enter (id) {
```
Comment (braddialpad) [inline packages/dialtone-vue3/components/hovercard/timer.js:15]: Maybe some short comment description, it's a bit hard to tell what is going on here without going through it line by line.

File: packages/dialtone-vue3/components/hovercard/timer.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+      }
+    } else {
+      timerEnter = setTimeout(() => {
+        current.value = id;
+        prevCard = id;
+      }, TOOLTIP_DELAY_MS);
+    }
+  }
+
+  function leave () {
```
Comment (braddialpad) [inline packages/dialtone-vue3/components/hovercard/timer.js:48]: Same here with comment

**Review by braddialpad — CHANGES_REQUESTED** [review]

File: packages/dialtone-vue2/components/lazy_show/lazy_show.vue (hunk 13 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/lazy_show/lazy_show.vue]: oh, that makes sense. Will probably have to find another ways of testing this, unless suspense works better with the automation. Thanks!

Comment (ninamarina) [reply packages/dialtone-vue2/components/lazy_show/lazy_show.vue]: I've been checking this out and even without the "initialized" logic, the component is not rendered the first time, because the show prop is still false. So I don't think this is actually making any performance difference.
Without the `v-if="initialized"` I can see that the DOM is only updated when the "show" property is true:
init-vue3

The slot from lazy-show is inside the div we see appearing and disappearing from the DOM in the gif:
image

I also tried this is vue 2 and it works the same way.

For what I can see, Suspense is for managing lazy loading of async dependencies, which is not what lazy show seems to be doing.

File: packages/dialtone-vue2/components/hovercard/hovercard.mdx (hunk 4 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.mdx]: Yes, will do

File: packages/dialtone-vue2/components/lazy_show/lazy_show.vue (hunk 13 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/lazy_show/lazy_show.vue]: Disregard my previous comment, the div is actually being rendered (with `display: none`), until the first time the anchor is hovered. It is inside the popover:
image

When doing the first hover, tippy moves this to the body element. So lazy-show is doing a render optimization.

The problem with the test was that I was expecting the elements to be inside the wrapper, while they are on the body.
This explains also what @iropolo pointed out about the tests, they are passing even without the timers because the div is being rendered only with `display: none`.

I will revert the changes to `lazy_show` and test that the element is present in the body element.

Comment (braddialpad) [reply packages/dialtone-vue2/components/lazy_show/lazy_show.vue]: Added ticket to look into the lazy-show further. https://dialpad.atlassian.net/browse/DLT-1477

once we get tests passing here, we're good to move forward.

File: packages/dialtone-vue3/components/hovercard/timer.js (hunk 14 above)
Comment (ninamarina) [reply packages/dialtone-vue3/components/hovercard/timer.js:15]: Fixed in latest changes, let me know if it's clear enough

File: packages/dialtone-vue3/components/hovercard/timer.js (hunk 15 above)
Comment (ninamarina) [reply packages/dialtone-vue3/components/hovercard/timer.js:48]: Fixed in latest changes

File: packages/dialtone-vue2/components/hovercard/hovercard.mdx (hunk 4 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.mdx]: Fixed in latest changes

File: packages/dialtone-vue2/components/hovercard/hovercard.mdx (hunk 5 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.mdx]: Fixed in latest changes

File: packages/dialtone-vue2/components/hovercard/hovercard.test.js (hunk 6 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.test.js]: Fixed in latest changes

File: packages/dialtone-vue2/components/hovercard/hovercard.vue (hunk 7 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.vue]: Fixed in latest changes

File: packages/dialtone-vue2/components/hovercard/hovercard.vue (hunk 8 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.vue]: Fixed in latest changes

File: packages/dialtone-vue2/components/hovercard/hovercard_constants.js
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard_constants.js:1]: Fixed in latest changes

File: packages/dialtone-vue2/components/hovercard/hovercard.test.js (hunk 12 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/hovercard/hovercard.test.js:60]: Fixed in latest changes

File: packages/dialtone-vue2/components/lazy_show/lazy_show.vue (hunk 13 above)
Comment (ninamarina) [reply packages/dialtone-vue2/components/lazy_show/lazy_show.vue]: Fixed in latest changes

File: packages/dialtone-vue3/components/hovercard/timer.js (hunk 15 above)
Comment (braddialpad) [reply packages/dialtone-vue3/components/hovercard/timer.js:48]: good description thanks.

**Review by braddialpad — APPROVED** [review]
Looks good by me, thanks!

**Comment by iropolo** [comment]
In vue2 version the hovercard close and open on click, this does not happen in vue3 version.
This could lead to an issue if you close the hovercard and mouseover again it quickly it does not appear again.

File: packages/dialtone-vue2/components/hovercard/hovercard.vue (hunk 7 above)
Comment (iropolo) [reply packages/dialtone-vue2/components/hovercard/hovercard.vue]: looks great

**Comment by iropolo** [comment]
Will suggest to move the template to the middle of the screen, like what we have on popover, in that way we can better see the `placement` prop working well.

**Comment by ninamarina** [comment]
> In vue2 version the hovercard close and open on click, this does not happen in vue3 version. This could lead to an issue if you close the hovercard and mouseover again it quickly it does not appear again.

Nice catch! Fixed in latest commit

**Review by iropolo — APPROVED** [review]
Looks great, nice job!!
